### ISSUE TITLE
Changing data in component with hook is not displayed on UI 

### ISSUE THREAD

**Issue by not-vainglorious** [strapi/strapi#13683]
### Required System information

- Node.js version: 16.14.2
- NPM version: 8.5.0
- Strapi version: 4.2.2
- Database: PostgreSQL 13  (Alpine)
- Operating system: Ubuntu 20.04.4

### Describe the bug

When updating component with beforeHook Strapi Admin UI is not changing value of number field. New value appears only after page refresh or interacting with field. (Put request is sending data)

### Steps to reproduce the behavior

1. Create collection/single
2. Create component with "Number" field
3. Add component to collection (tested as single component)
4. Add before update hook that somehow changes field value with entity service update (code will be below) 
5. Save and check network / field

### Expected behavior

Field value should refresh as it working with other field types inside component.

### Screenshots

After clicking save

![Screenshot from 2022-07-05 03-12-43](https://user-images.githubusercontent.com/88206744/177227659-feec112b-d19b-405a-8667-e665e1ac9a3c.png)

Field focus

![Screenshot from 2022-07-05 03-12-48](https://user-images.githubusercontent.com/88206744/177227701-23f99177-3a3a-4754-83b9-a59474934d25.png)

Remove focus

![Screenshot from 2022-07-05 03-12-51](https://user-images.githubusercontent.com/88206744/177227739-cff49a31-b986-4234-b162-091cf250cb8d.png)

### Code snippets

```
module.exports = {
  beforeUpdate: async (event) => {
    console.log(event.params.data.test);

    if (event.params.data.test.id) {
      await strapi.entityService.update(
        'test.test',
        event.params.data.test.id,
        {
          data: { text: '1', number: 1, boolean: true },
        },
      );
    }
  },
};
```

### Additional context

Request:
```
{
      createdAt: "2022-07-04T23:58:43.896Z"
      createdBy: 1
      id: 1
      publishedAt: null
      sitemap_exclude: false
      test: {}
      updateField: "123345"
      updatedAt: "2022-07-05T00:10:28.626Z"
      updatedBy: 1
}
```
Response:
```
{
"id":1,
      "createdAt":"2022-07-04T23:58:43.896Z",
      "updatedAt":"2022-07-05T00:10:33.493Z",
      "publishedAt":null,
      "sitemap_exclude":false,
      "updateField":"123345",
      "test":{"id":4,"text":"1","number":1,"boolean":true}
}
```

**Comment by kasonde** [comment]
Looks like this only happens for number fields.

**Comment by not-vainglorious** [comment]
> Looks like this only happens for number fields.

Yep. This one is possible due to number field value from props which is passing to components state but doesn't handle props update correctly. If there will be any new  field type which generates state with usage of value you should always handle props change so new values that will come from props will be correctly updated.
